src/sink.rs

```diff
@@ -125,24 +170,61 @@ where
         self.pipelines.insert(id, addr);
     }
 
+    fn add_dest_pipeline(&mut self, id: TremorURL, addr: pipeline::Addr) {
+        self.dest_pipelines.push((id, addr));
+    }
+
     fn remove_pipeline(&mut self, id: TremorURL) -> bool {
         self.pipelines.remove(&id);
         self.pipelines.is_empty()
     }
 
     async fn on_signal(&mut self, signal: Event) -> Option<Event> {
-        let insights = self.sink.on_signal(signal).await.ok()??;
-        for insight in insights {
-            let mut i = self.pipelines.values_mut();
-            if let Some(first) = i.next() {
-                for p in i {
-                    if let Err(e) = p.send_insight(insight.clone()).await {
-                        error!("Error: {}", e)
-                    };
+        let replies = self.sink.on_signal(signal).await.ok()??;
+        for reply in replies {
+            match reply {
+                SinkReply::Insight(e) => {
+                    let mut i = self.pipelines.values_mut();
+                    if let Some(first) = i.next() {
+                        for p in i {
+                            if let Err(e) = p.send_insight(e.clone()).await {
+                                error!("Error: {}", e)
+                            };
+                        }
+                        if let Err(e) = first.send_insight(e).await {
+                            error!("Error: {}", e)
+                        };
+                    }
+                }
+                SinkReply::Response(e) => {
```

Comment:
handling the response here can be problematic. this is only ever executed as a result of a signal. right now that would introduce a 100ms delay to any reply.